File: cases/taint/javascript/recursive-callback-transform-negative/RecursiveCallbackTransformNegative.js
function dfb_source() { // DFB-SOURCE: recursive-callback-transform-input
  return 7;
}

function dfb_sink(value) { } // DFB-SINK: recursive-callback-transform-sink

function walk(value, depth, step) {
  if (depth === 0) {
    value = 0; // DFB-KILL: recursive-callback-transform-base-overwrite
    return value; // DFB-WITNESS: recursive-callback-transform-base
  }
  return step(value, depth - 1); // DFB-WITNESS: recursive-callback-transform-transfer
}

function step(value, depth) {
  const nextValue = value + 1; // DFB-WITNESS: recursive-callback-transform-step
  const recursiveResult = walk(nextValue, depth, step); // DFB-WITNESS: recursive-callback-transform-recursive-transfer
  return recursiveResult + 1; // DFB-WITNESS: recursive-callback-transform-compose
}

function run() {
  dfb_sink(walk(dfb_source(), 3, step));
}

File: cases/taint/ruby/recursive-callback-transform-negative/recursive_callback_transform_negative.rb
def dfb_source # DFB-SOURCE: recursive-callback-transform-input
  return 7
end

def dfb_sink(value) # DFB-SINK: recursive-callback-transform-sink
end

def walk(value, depth, callback)
  if depth == 0
    value = 0 # DFB-KILL: recursive-callback-transform-base-clean
    return value # DFB-WITNESS: recursive-callback-transform-base
  end
  callback.call(value, depth - 1) # DFB-WITNESS: recursive-callback-transform-indirect-transfer
end

def step(value, depth)
  next_value = value + 1 # DFB-WITNESS: recursive-callback-transform-step
  recursive = walk(next_value, depth, method(:step)) # DFB-WITNESS: recursive-callback-transform-recursive-transfer
  recursive + 1 # DFB-WITNESS: recursive-callback-transform-compose
end

def run
  dfb_sink(walk(dfb_source, 3, method(:step)))
end
